waspc/waspls/src/Wasp/LSP/SignatureHelp.hs

```diff
@@ -0,0 +1,261 @@
+        let fields = intersperse ",\n  " (map fieldToFragment (M.toList fieldMap))
+         in concat [["{\n  "], fields, ["\n}"]]
+  Type.ListType inner -> ["[", Param List (showInnerType inner), "]"]
+  Type.TupleType (a, b, cs) ->
+    let fieldTypes = a : b : cs
+        fields = intersperse ", " (zipWith (\n -> Param (Tuple n) . showInnerType) [0 ..] fieldTypes)
+     in concat [["("], fields, [")"]]
+  typ -> [fromString (show typ)]
+  where
+    showInnerType :: IsString s => Type -> s
+    showInnerType (Type.DictType _) = "{ ... }"
+    showInnerType typ = fromString (show typ)
+
+    fieldToFragment :: (String, Type.DictEntryType) -> SignatureFragment
+    fieldToFragment (key, Type.DictRequired typ) =
+      Param (Key key) $ printf "%s: %s" key (showInnerType typ :: String)
+    fieldToFragment (key, Type.DictOptional typ) =
+      Param (Key key) $ printf "%s?: %s" key (showInnerType typ :: String)
+
+-- | Convert a list of fragments to a string by concatenating the text of each
+-- fragment.
+showFragments :: [SignatureFragment] -> String
+showFragments = concatMap fragmentText
+
+-- | Search through a list of fragments to find the parameters.
+--
+-- The LSP wants the starting and ending offset of each parameter in the signature
+-- help text (computed in 'showFragments'), so this function also has to keep
+-- track of the 0-based offset into the final text.
+getParamsFromFragments :: [SignatureFragment] -> [LSP.ParameterInformation]
+getParamsFromFragments fragments =
+  reverse $ map labelToInfo $ snd $ foldl' go (0, []) fragments
+  where
+    -- A left fold is used because offset needs to increase as you go right across
+    -- the list. This means the parameter list needs to be reversed after the fold
+    -- so that they are in the same order as the parameters in the fragments.
+    go :: (LSP.UInt, [LSP.ParameterLabel]) -> SignatureFragment -> (LSP.UInt, [LSP.ParameterLabel])
+    go (offset, labels) (Text text) = (offset + fromIntegral (length text), labels)
+    go (offset, labels) (Param _ text) =
+      let end = offset + fromIntegral (length text)
+       in (end, LSP.ParameterLabelOffset offset end : labels)
+
+    labelToInfo :: LSP.ParameterLabel -> LSP.ParameterInformation
+    labelToInfo label =
+      LSP.ParameterInformation
+        { _label = label,
+          _documentation = Nothing
+        }
+
+-- | Find the index of the active parameter in a list of fragments.
+--
+-- NOTE: This function computes the index into the parameter list, not the
+-- fragment list. This is why plain 'Text' fragments are filtered out of the
+-- list before indexing (via @mapMaybe fragmentParam@).
+findActiveParameterIndex :: [SignatureFragment] -> ExprKey -> Maybe LSP.UInt
+findActiveParameterIndex fragments key =
+  fromIntegral <$> elemIndex key (mapMaybe fragmentParam fragments)
+
+-- | Lift a 'Maybe' into a 'MaybeT' monad transformer.
+hoistMaybe :: Monad m => Maybe a -> MaybeT m a
+hoistMaybe = MaybeT . pure
```

Comment:
Uf I am finding it a bit hard to understand this, could you help me a bit?

So `MaybeT m a` is a monad transformer stack, therefore a monad that has special powers of `Maybe` on the outside, but has powers of monad `m` on the inside. So it is a combo of `Maybe` and some `m` monad.

Now this function takes a `Maybe a` and returns `MaybeT m a`.
Ok, so the value `a` remains the same, sure, we just change the "wrapper" around it.

So we do `MaybeT . pure`.

Signature of pure is `pure :: a -> m a`, and it just wraps anything into a any monad.

Signature of `MaybeT` is `MaybeT :: m a -> MaybeT m a`. Ah no, that is wrong -> ok that is what confused me. So signature is `MaybeT :: m (Maybe a) -> MaybeT m a`. So basically it takes a "normal" combo of Maybe and some monad and turns it into a transformer, a powered up monad.

Ok, makes then, all the types fit of course.

So I really take `Maybe` value and pump it up to also have these aditional powers from this `m` monad.

But why `hoist` then? Why not `liftMaybe`? Does that already have some other meaning reserved? what about `maybeToMaybeT`? Or `liftToMaybeT`?